internal/hook/wait_exec_hook_handler.go

```diff
@@ -0,0 +1,260 @@
+				cancel()
+				return
+			}
+
+			// Sequentially run all hooks for the ready container. The container's hooks are not
+			// removed from the byContainer map until all have completed so that if one fails
+			// remaining unexecuted hooks can be handled by the outer function.
+			for i, hook := range hooks {
+				// This indicates to the outer function not to handle this hook as unexecuted in
+				// case of terminating before deleting this container's slice of hooks from the
+				// byContainer map.
+				byContainer[containerName][i].executed = true
+
+				hookLog := log.WithFields(
+					logrus.Fields{
+						"hookSource": hook.Source,
+						"hookType":   "exec",
+						"hookPhase":  "post",
+					},
+				)
+				// Check the individual hook's wait timeout is not expired
+				if hook.Hook.WaitTimeout.Duration != 0 && time.Since(waitStart) > hook.Hook.WaitTimeout.Duration {
+					err := fmt.Errorf("Hook %s in container %s in pod %s/%s expired before executing", hook.Name, hook.Hook.Container, pod.Namespace, pod.Name)
+					hookLog.Error(err)
+					if hook.Hook.OnError == velerov1api.HookErrorModeFail {
+						errors = append(errors, err)
+						cancel()
+						return
+					}
+				}
+				eh := &velerov1api.ExecHook{
+					Container: hook.Hook.Container,
+					Command:   hook.Hook.Command,
+					OnError:   hook.Hook.OnError,
+					Timeout:   hook.Hook.ExecTimeout,
+				}
+				if err := e.PodCommandExecutor.ExecutePodCommand(hookLog, podMap, pod.Namespace, pod.Name, hook.Name, eh); err != nil {
+					hookLog.WithError(err).Error("Error executing hook")
+					if hook.Hook.OnError == velerov1api.HookErrorModeFail {
+						errors = append(errors, err)
+						cancel()
+						return
+					}
+				}
+			}
+			delete(byContainer, containerName)
+		}
+		if len(byContainer) == 0 {
+			cancel()
+		}
+	}
+
+	selector := fields.OneTermEqualSelector("metadata.name", pod.Name)
+	lw := e.ListWatchFactory.NewListWatch(pod.Namespace, selector)
+
+	_, controller := cache.NewInformer(lw, pod, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: handler,
+		UpdateFunc: func(_, newObj interface{}) {
+			handler(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			err := fmt.Errorf("Pod %s/%s deleted before all hooks were executed", pod.Namespace, pod.Name)
+			log.Error(err)
+			cancel()
+		},
+	})
+
+	controller.Run(ctx.Done())
+
+	// There are some cases where this function could return with unexecuted hooks: the pod may
+	// be deleted, a hook with OnError mode Fail could fail, or it may timeout waiting for
+	// containers to become ready.
+	// Each unexecuted hook is logged as an error but only hooks with OnError mode Fail return
+	// an error from this function.
+	for _, hooks := range byContainer {
+		for _, hook := range hooks {
+			if hook.executed {
+				continue
+			}
+			err := fmt.Errorf("Hook %s in container %s in pod %s/%s not executed", hook.Name, hook.Hook.Container, pod.Namespace, pod.Name)
```

Comment:
can we also capture why we didn't run the hook?

Reply:
I append the context error, but that's just `context deadline exceeded`. It would be possible to know the reason if something in this function canceled the context, but it could also be canceled by something outside.

Reply:
Thanks for clarifying!